server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -410,4 +414,22 @@ private static void ensureNoAwarenessAttributeDecommissioned(
             );
         }
     }
+
+    private static void ensureNoQuorumLossDueToDecommissioning(
+        DecommissionAttribute decommissionAttribute,
+        Set<DiscoveryNode> clusterManagerNodesToBeDecommissioned,
+        CoordinationMetadata.VotingConfiguration votingConfiguration
+    ) {
+        final Set<String> nodesInVotingConfig = votingConfiguration.getNodeIds();
+        assert nodesInVotingConfig.isEmpty() == false;
+        final int requiredNodes = nodesInVotingConfig.size() / 2 + 1;
+        final Set<String> realNodeIds = new HashSet<>(nodesInVotingConfig);
+        realNodeIds.removeIf(ClusterBootstrapService::isBootstrapPlaceholder);
+        if (realNodeIds.size() - clusterManagerNodesToBeDecommissioned.size() < requiredNodes) {
+            throw new DecommissioningFailedException(
+                decommissionAttribute,
+                "cannot proceed with decommission request. Cluster might go into quorum loss"
+            );
+        }
```

Comment:
should we do explicit node id based checks by checking the intersection?

Reply:
Sure. Updating logic to - 
```
Set<String> clusterManagerNodesIdToBeDecommissioned = new HashSet<>();
        clusterManagerNodesToBeDecommissioned.forEach(node -> clusterManagerNodesIdToBeDecommissioned.add(node.getId()));
        if (!votingConfiguration.hasQuorum(
            votingConfiguration.getNodeIds()
                .stream()
                .filter(n -> clusterManagerNodesIdToBeDecommissioned.contains(n) == false)
                .collect(Collectors.toList()
                )
            )
        ) {
            throw new DecommissioningFailedException(
                decommissionAttribute,
                "cannot proceed with decommission request as cluster might go into quorum loss"
            );
        }
```